provider/cmd/crd2pulumi/generate.go

```diff
@@ -0,0 +1,209 @@
+	group, _, _ := unstruct.NestedString(crd.Object, "spec", "group")
+
+	if outputPath == "" {
+		defaultOutputPath := func(yamlPath, plural, language string) string {
+			var extension string
+			switch language {
+			case NodeJS:
+				extension = "ts"
+			case DotNet:
+				extension = "cs"
+			case Python:
+				extension = "py"
+			case Go:
+				extension = "go"
+			}
+			outputFileName := plural + "." + extension
+			return path.Join(filepath.Dir(yamlPath), outputFileName)
+		}
+		outputPath = defaultOutputPath(yamlPath, plural, language)
+	}
+
+	customResourceGenerator := CustomResourceGenerator{
+		CustomResourceDefinition: crd,
+		OutputPath:               outputPath,
+		Language:                 language,
+		APIVersion:               apiVersion,
+		Kind:                     kind,
+		Plural:                   plural,
+		Group:                    group,
+		Versions:                 versions,
+	}
+	return customResourceGenerator, nil
+}
+
+func (gen *CustomResourceGenerator) GetName() string {
+	return gen.Plural + "." + gen.Group
+}
+
+// GetVersionNames returns a slice of the versions supported by this CRD.
+func (gen *CustomResourceGenerator) GetVersionNames() []string {
+	versionNames := make([]string, 0, len(gen.Versions))
+	for versionName := range gen.Versions {
+		versionNames = append(versionNames, versionName)
+	}
+	return versionNames
+}
+
+func (gen *CustomResourceGenerator) GenerateCode() error {
+	objectTypeSpecs := GetObjectTypeSpecs(gen.Versions, gen.GetName(), gen.Kind)
+	PrettyPrint(objectTypeSpecs)
+
+	switch gen.Language {
+	case NodeJS:
+		types, err := nodejs.GenerateTypes(objectTypeSpecs)
+		if err != nil {
+			return fmt.Errorf("generate code: %v", err)
+		}
+		classes := gen.GenerateNodeJSClasses()
+
+		file, err := os.Create(gen.OutputPath)
+		if err != nil {
+			return fmt.Errorf("creating file at %s: %v", gen.OutputPath, err)
+		}
+		file.WriteString(types)
+		file.WriteString(classes)
+		defer file.Close()
+		return nil
+	case DotNet:
+	case Python:
+	case Go:
+	}
+
+	return nil
+}
+
+func (gen *CustomResourceGenerator) GenerateNodeJSClasses() string {
+	versionNames := gen.GetVersionNames()
+	kind := gen.Kind
+	group := gen.Group
+	name := gen.GetName()
+
+	// Generates a CustomResource sub-class for a single version
+	generateResourceClass := func(sb *strings.Builder, version, kind, name, group string) {
+		argsName := version + "." + kind
+		apiVersion := fmt.Sprintf("%s/%s", group, version)
+		sb.WriteString(fmt.Sprintf("export namespace %s {\n", version))
+		sb.WriteString(fmt.Sprintf("\texport class %s extends k8s.apiextensions.CustomResource {\n", kind))
+		sb.WriteString(fmt.Sprintf("\t\tconstructor(name: string, args?: %s, opts?: pulumi.CustomResourceOptions) {\n", argsName))
+		sb.WriteString(fmt.Sprintf("\t\t\tsuper(name, { apiVersion: \"%s\", kind: \"%s\", ...args }, opts)\n", apiVersion, kind))
+		sb.WriteString(fmt.Sprintf("\t\t}\n\t}\n}\n\n"))
+	}
```

Comment:
Alternatively to using `*strings.Builder`, you could do something like:

```suggestion
	generateResourceClass := func(w io.Writer, version, kind, name, group string) {
		argsName := version + "." + kind
		apiVersion := fmt.Sprintf("%s/%s", group, version)
		fmt.Fprintf(w, "export namespace %s {\n", version)
		fmt.Fprintf(w, "\texport class %s extends k8s.apiextensions.CustomResource {\n", kind)
		fmt.Fprintf(w, "\t\tconstructor(name: string, args?: %s, opts?: pulumi.CustomResourceOptions) {\n", argsName)
		fmt.Fprintf(w, "\t\t\tsuper(name, { apiVersion: \"%s\", kind: \"%s\", ...args }, opts)\n", apiVersion, kind)
		fmt.Fprintf(w, "\t\t}\n\t}\n}\n\n")
	}
```


And then:

```go
	buffer := &bytes.Buffer{}
 	for _, version := range versionNames {
 		generateResourceClass(buffer, version, kind, name, group)
 	}
 	generateDefinitionClass(buffer, kind, gen.APIVersion)

 	return buffer.String()
```

Reply:
Is there any performance advantage with using buffers compared to StringBuilders? Or is it just simpler since we can directly use string formatting in the `fmt.Fprintf` function?

Reply:
Just for the simpler `fmt.Fprintf`. I'm not sure about the perf difference between `strings.Builder` vs. `bytes.Buffer`, but I doubt it makes a measurable difference here. It looks like `strings.Builder` also implements the `io.Writer` interface, so it could be used with `fmt.Fprintf`.